### PR TITLE
Enable co function optimization by V8

### PR DESCRIPTION
**Pull request by ruimarinho**
Inlining slicing of arguments allows V8 to optimize the main `co` function. Also added a tiny test which was missing when `co` receives is called with no arguments.

**Before**:

```
[compiling method 0x15a40d13ab61 <JS Function co (SharedFunctionInfo 0x2d5161561089)> using Crankshaft]
[aborted optimizing 0x15a40d13ab61 <JS Function co (SharedFunctionInfo 0x2d5161561089)> because: Bad value context for arguments value]
```

**After**:

```
[compiling method 0x3a75df13acf9 <JS Function co (SharedFunctionInfo 0x3f32caa61089)> using Crankshaft]
[optimizing 0x3a75df13acf9 <JS Function co (SharedFunctionInfo 0x3f32caa61089)> - took 0.096, 0.130, 0.100 ms]
```

This is just a small optimisation but we have to start somewhere :)

### CODE REVIEW

**Comment by cesarandreu** [comment]
@ruimarinho How do you get this info? What does each of the numbers `took 0.096, 0.130, 0.100 ms ` mean?

**Comment by bucaran** [comment]
@ruimarinho IMO this hurts readability more than it optimizes performance, but on the other hand, `co` is a low level library and if this could potentially benefit higher level libraries I think it's acceptable.

Having said that, how often do you think people will use this new feature, I mean, `co` had been around years before I [introduced that](https://github.com/tj/co/pull/232).

**Comment by ruimarinho** [comment]
@cesarandreu you can use the `--trace-deopt` flag to get this info. That exact timing is not important, just the "aborted optimising" vs "optimising" parts. 

@bucaran completely agree, it's a trade off. In fact, I should clarify two things - this PR is not related to your changes. I just reminded of an optimisation that should have also been present originally on a [different function](https://github.com/tj/co/pull/235/files#diff-168726dbe96b3ce427e7fedce31bb0bcL138), so I took the time to change them both. Secondly, the usage of the feature has no impact on V8 - the single fact the `arguments` object is touched using `slice.call` will deoptimize the function. It is also not just a question of performance - memory is impacted too, since on both situations `arguments` would cause leak.

**Comment by bucaran** [comment]
@ruimarinho I see. Very interesting stuff. It would be fantastic if you could provide a bit more info on the subject, or provide links so that it's clear to everyone. I, myself, don't know much about V8 optimizations.

:+1:

**Comment by madbence** [comment]
@bucaran [Optimalization killers](https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#3-managing-arguments)

**Comment by bucaran** [comment]
@madbence Awesome, thanks.

File: index.js [hunk 1]
```diff
@@ -42,7 +36,10 @@ co.wrap = function (fn) {
 
 function co(gen) {
   var ctx = this;
-  var args = slice.call(arguments, 1)
+  var args = arguments.length === 0 ? [] : new Array(arguments.length - 1);
+  for (var $i = 0; $i < args.length; ++$i) {
```
Comment (fengmk2) [inline index.js]: `$i` => `i` will be better.

**Comment by fengmk2** [comment]
+1 for optimization.

/cc @tj @jonathanong

**Comment by tj** [comment]
I'd only bother with stuff like this if it has real world benefits, people have a tendency to micro-optimize in js land for no reason haha

**Comment by arusakov** [comment]
@fengmk2 @tj 
Looks at this [commit](https://github.com/nodejs/node-v0.x-archive/commit/91f1b250ecb4fb8151cd17423dd4460652d0ce97) in Node.js. It was a long time ago. It's micro-optinization maybe, maybe not for V8. However, this optimization is still important for V8. For example, I tried to remove ```slice(arguments)``` in [PR for sockjs-client](https://github.com/sockjs/sockjs-client/pull/263) library and write some [benchmark](https://github.com/sockjs/sockjs-client/pull/263#issuecomment-144563735) (5x speed up for Node.js v4). Replace ```slice(arguments)``` to manual copying to new array is really work.

**Comment by tejasmanohar** [comment]
I'm +1. sure it's a micro optimization but it's an optimization without harm. i wouldn't put time into it myself but since it's already solved here

**Comment by tj** [comment]
Happy to consider if we get some real-world numbers that aren't micro-optimizations

File: index.js (hunk 1 above)
Comment (luckydrq) [inline index.js]: haha, same old optimization

**Pull request closed without merging**